src/client/csfle/client_encryption.rs

```diff
@@ -0,0 +1,354 @@
+                        "$$REMOVE",
+                        {
+                            "$filter": {
+                                "input": "$keyAltNames",
+                                "cond": { "$ne": ["$$this", key_alt_name] },
+                            }
+                        }
+                    ]
+                }
+            }
+        };
+        self.key_vault
+            .find_one_and_update(doc! { "_id": id }, vec![update], None)
+            .await
+    }
+
+    /// Returns a key document in the key vault collection with the given keyAltName.
+    pub async fn get_key_by_alt_name(&self, key_alt_name: &str) -> Result<Option<RawDocumentBuf>> {
+        self.key_vault
+            .find_one(doc! { "keyAltNames": key_alt_name }, None)
+            .await
+    }
+
+    /// Encrypts a BsonValue with a given key and algorithm.
+    /// Returns an encrypted value (BSON binary of subtype 6).
+    pub async fn encrypt(&self, value: bson::RawBson, opts: EncryptOptions) -> Result<Binary> {
+        let ctx = self.encrypt_ctx(value, opts)?;
+        let result = self.exec.run_ctx(ctx, None).await?;
+        let bin_ref = result
+            .get_binary("v")
+            .map_err(|e| Error::internal(format!("invalid encryption result: {}", e)))?;
+        Ok(bin_owned(bin_ref))
+    }
+
+    fn encrypt_ctx(&self, value: bson::RawBson, opts: EncryptOptions) -> Result<Ctx> {
+        let mut builder = self.crypt.ctx_builder();
+        match opts.key {
+            EncryptKey::Id(id) => {
+                builder = builder.key_id(&id.bytes)?;
+            }
+            EncryptKey::AltName(name) => {
+                builder = builder.key_alt_name(&name)?;
+            }
+        }
+        builder = builder.algorithm(opts.algorithm)?;
+        if let Some(factor) = opts.contention_factor {
+            builder = builder.contention_factor(factor)?;
+        }
+        if let Some(qtype) = opts.query_type {
+            builder = builder.query_type(&qtype)?;
+        }
+        Ok(builder.build_explicit_encrypt(value)?)
+    }
+
+    /// Decrypts an encrypted value (BSON binary of subtype 6).
+    /// Returns the original BSON value.
+    pub async fn decrypt<'a>(&self, value: RawBinaryRef<'a>) -> Result<bson::RawBson> {
+        if value.subtype != BinarySubtype::Encrypted {
+            return Err(Error::invalid_argument(format!(
+                "Invalid binary subtype for decrypt: expected {:?}, got {:?}",
+                BinarySubtype::Encrypted,
+                value.subtype
+            )));
+        }
+        let ctx = self
+            .crypt
+            .ctx_builder()
+            .build_explicit_decrypt(value.bytes)?;
+        let result = self.exec.run_ctx(ctx, None).await?;
+        Ok(result
+            .get("v")?
+            .ok_or_else(|| Error::internal("invalid decryption result"))?
+            .to_raw_bson())
+    }
+}
+
+/// Options for initializing a new `ClientEncryption`.
+#[derive(TypedBuilder)]
```

Comment:
I think it might be worth going over these various structs to see which ones should be `Clone` (unless they're intentionally not so already).

Reply:
Good thought, went through and added `Clone` and `Debug` derives.